mlflow/pytorch/_pytorch_autolog.py

```diff
@@ -0,0 +1,206 @@
+# In case of MlflowLogger, Run management is completely controlled by the class and
+# hence mlflow object needs to be reinstantiated by setting
+# tracking uri, experiment_id and run_id which may lead to a race condition.
+
+
+@rank_zero_only
+def _autolog(log_every_n_iter=1):
+    global every_n_iter
+    every_n_iter = log_every_n_iter
+
+    class __MLflowPLCallback(pl.Callback):
+        """
+        Callback for auto-logging metrics and parameters.
+        """
+
+        def __init__(self):
+            self.early_stopping = False
+
+        def on_epoch_end(self, trainer, pl_module):
+            """
+            Log loss and other metrics values after each epoch
+
+            :param trainer: pytorch lightning trainer instance
+            :param pl_module: pytorch lightning base module
+            """
+            if (pl_module.current_epoch + 1) % every_n_iter == 0:
+                for key, value in trainer.callback_metrics.items():
+                    try_mlflow_log(
+                        mlflow.log_metric, key, float(value), step=pl_module.current_epoch
+                    )
+
+            for callback in trainer.callbacks:
+                if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                    self._early_stop_check(callback)
+
+        def on_train_start(self, trainer, pl_module):
+            """
+            Logs Optimizer related metrics when the train begins
+
+            :param trainer: pytorch lightning trainer instance
+            :param pl_module: pytorch lightning base module
+            """
+            try_mlflow_log(mlflow.set_tag, "Mode", "training")
+            try_mlflow_log(mlflow.log_param, "epochs", trainer.max_epochs)
+
+            for callback in trainer.callbacks:
+                if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                    self.early_stopping = True
+                    self._log_early_stop_params(callback)
+
+            if hasattr(trainer, "optimizers"):
+                for optimizer in trainer.optimizers:
+                    try_mlflow_log(mlflow.log_param, "optimizer_name", type(optimizer).__name__)
+
+                    if hasattr(optimizer, "defaults"):
+                        defaults_dict = optimizer.defaults
+                        if "lr" in defaults_dict:
+                            try_mlflow_log(mlflow.log_param, "learning_rate", defaults_dict["lr"])
+
+                        if "eps" in defaults_dict:
+                            try_mlflow_log(mlflow.log_param, "epsilon", defaults_dict["eps"])
+
+                        if "betas" in defaults_dict:
+                            try_mlflow_log(mlflow.log_param, "betas", defaults_dict["betas"])
+
+                        if "weight_decay" in defaults_dict:
+                            try_mlflow_log(
+                                mlflow.log_param, "weight_decay", defaults_dict["weight_decay"]
+                            )
```

Comment:
@shrinath-suresh cc @smurching 

Not sure if we need to address this in this PR, but we should consider how to handle the case where multiplte optimizers are given.

Reply:
@harupy yeah good catch, currently it seems possible to log e.g. the learning rate from one optimizer and the epsilon value from another optimizer, depending on which optimizers contain which fields. @shrinath-suresh would it make sense to suffix the parameter & metric names with the optimizer name in the case of more than one optimizer? e.g. something like `learning_rate_adam_optimizer`?

Reply:
or prefix with a class name (e.g. `Adam.learning_rate`)?

![Screen Shot 2020-11-02 at 12 29 21](https://user-images.githubusercontent.com/17039389/97827283-25bc3500-1d07-11eb-8aba-c1d1876997ef.png)

Reply:
But this approach doesn't work when the users uses two Adam optimizers (not sure how realistic it is)...

Reply:
@harupy including the class name in the metric/param name SGTM. I'd actually suggested suffixing instead of prefixing because the type of metric (learning rate, weight decay, etc) seemed more important than the optimizer it comes from, and we truncate metric names e.g. in the experiment runs table UI, as in this screenshot:

![image](https://user-images.githubusercontent.com/2358483/97933886-e1419100-1d28-11eb-9ed9-058d83e7f6ba.png)

That is, it'd be nice to see a truncated metric name like "learning_rate..." instead of "Adam.lear..."

Reply:
From offline discussion -  if we want to accelerate getting this PR in, we could simplify the metric/param logging logic to only log metrics/params of the first optimizer, then in a follow-up add support for multiple optimizers - I'm guessing it's not common to have multiple optimizers in general

Reply:
@smurching  https://github.com/mlflow/mlflow/pull/3601/commits/f3f118b45123d9daa1125fed8c88b711f384a5fa - we have already addressed this as part of this commit. We loop through the optimizers and save the parameters as `{param}_{optimizer_name}`. For Ex: `learning_rate_adam_optimizer`